### PR TITLE
Add access condition models to Azure Core

### PR DESCRIPTION
**Pull request by alzimmermsft**
Implements #5938 

Adds `MatchConditions` and `RequestConditions` model classes to Azure Core to support HTTP access conditions.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+// Licensed under the MIT License.
+
+package com.azure.core.http;
+
+import java.time.OffsetDateTime;
+
+/**
+ * Specifies HTTP options for conditional requests based on modification time.
+ */
+public class RequestConditions extends MatchConditions {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java]: Add @Fluent annotation.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    @Override
+    public RequestConditions setIfMatch(String ifMatch) {
+        super.setIfMatch(ifMatch);
+        return this;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return The updated ResourceConditions object.
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java]: {@inheritDoc} + new doc doesn't generate the Javadocs correctly. Use `@inheritDoc` only or add new javadoc.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java]: Done

File: sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java]: Pulled the super class documentation class directly.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/MatchConditions.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (conniey) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/MatchConditions.java]: @mssfang Would this make your azconfig code easier?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+     *
+     * @return The datetime that resources must have remained unmodified since.
+     */
+    public OffsetDateTime getIfUnmodifiedSince() {
+        return ifUnmodifiedSince;
+    }
+
+    /**
+     * Optionally limit requests to resources that have remained unmodified since the passed
+     * {@link OffsetDateTime datetime}.
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/RequestConditions.java]: There is one :)

File: sdk/core/azure-core/src/main/java/com/azure/core/http/MatchConditions.java (hunk 3 above)
Comment (mssfang) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/MatchConditions.java]: That makes AzConfig to have more public APIs to maintain. 🔢

**Pull request merged**